cpp/src/arrow/device.h

```diff
@@ -165,6 +215,20 @@ class ARROW_EXPORT MemoryManager : public std::enable_shared_from_this<MemoryMan
   static Result<std::shared_ptr<Buffer>> ViewBuffer(
       const std::shared_ptr<Buffer>& source, const std::shared_ptr<MemoryManager>& to);
 
+  /// \brief Create a SyncEvent for exporting
+  ///
+  /// This version should construct the appropriate event for the device and
+  /// provide the unique_ptr with the correct deleter for the event type.
+  virtual Result<std::shared_ptr<Device::SyncEvent>> MakeDeviceSyncEvent();
+
+  /// \brief Create a SyncEvent from imported device array.
+  ///
+  /// @param sync_event passed in sync_event from the imported device array.
+  /// @param release_sync_event destructor to free sync_event. `nullptr` may be
+  ///        passed to indicate that no destruction/freeing is necessary
+  virtual Result<std::shared_ptr<Device::SyncEvent>> MakeDeviceSyncEvent(
```

Comment:
Call this `WrapDeviceSyncEvent` to stress that we're just reusing an existing one?
I'm also surprised that the `release_fn_t` needs to be given explicitly here. Presumably, each MemoryManager implementation will know about a single kind of sync event?

Reply:
Hmm, so there are two possible uses: wrap an owned sync event and wrap a borrowed one, is it right?
If that's the case, then passing the actual release func, isn't necessary. Just need a boolean flag, or two separate methods for clarity.

Reply:
Or perhaps I'm misunderstanding and this is used only when importing from the C Data Interface... :-D

Reply:
Currently this is only being used when importing from the C Data Interface. I think we kept the release_sync_event param to keep this as generic as possible

Reply:
I agree that we need a way to recreate a managed SyncEvent from imported C Device Data. However, it could also be a private `ImportedSyncEvent` class in the C bridge implementation.

Reply:
It would still need to be created by the device / memory manager in order to create the correct derived type for handling `Wait` / `WaitEvent` etc.